Group block: use a variation picker in the placeholder

ramonjd opened the pull request:
## What?
Implement a variation-selector placeholder for the Group block for web browsers (not mobile).

Resolves: https://github.com/WordPress/gutenberg/issues/43433

## Why?
For ease of access to current and future layouts.

## How?
Using `<__experimentalBlockVariationPicker />` also used by the [Columns block](https://github.com/WordPress/gutenberg/blob/29d6801902ad97f749ccadd789a23b2d97369e3e/packages/block-library/src/columns/edit.js#L226-L225).

## Testing Instructions
In trunk create a few group blocks:

<details>

<summary>Here are some:</summary>

```html
<!-- wp:group {"layout":{"type":"constrained","isDefault":true}} -->
<div class="wp-block-group"></div>
<!-- /wp:group -->

<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:group {"layout":{"type":"constrained","isDefault":true}} -->
<div class="wp-block-group"></div>
<!-- /wp:group --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:group {"layout":{"type":"constrained","isDefault":true}} -->
<div class="wp-block-group"></div>
<!-- /wp:group --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

```
</details>

Save the post.

Checkout this branch `git checkout -t origin/try/group-variation-placeholder`.

Refresh the editor.

Check that the Group blocks appear as they should (no placeholder).

Now insert a new Group block.

You should see the variation placeholder picker.

Check behaviour in the site editor as well.


### Expectations:

- When a flex layout (row or stack) is selected, the placeholder should not be displayed. Even after refreshing the page.
- When the default group variation has been selected, and styles are applied, the placeholder should not be displayed. Even after refreshing the page.
- When there are innerblocks, the placeholder should not be displayed. Even after refreshing the page.
- The layout should default to `is-layout-constrained` when transforming blocks to Group blocks, including multiselect transforms.
- The group block should work as expected in block and classic themes.
- After selecting a group variation the newly-inserted block should be selected.


## Screenshots or screencast <!-- if applicable -->


<img width="902" alt="Screen Shot 2022-11-04 at 12 33 23 pm" src="https://user-images.githubusercontent.com/6458278/199866156-a772771f-3759-47f6-a3d0-1592adc4bc16.png">

ramonjd committed:
Group block: remove default layout for the group block and use a variation picker in the placeholder.

ramonjd committed:
Updating e2e tests to select default group from group variation placeholder

ramonjd committed:
Adding meaningful aria label to block variation picker
Updating e2e tests
Refactoring Group edit.js

ramonjd committed:
Only you..... can make my tests go green....

ramonjd committed:
Clicking on block variation skip button

ramonjd commented:
To follow up:

- Check whether the templates need a layout`<!-- wp:group {"templateLock":false,"layout":{"type":"constrained"}} -->`. See failing snapshot: https://github.com/WordPress/gutenberg/blob/93e29bbe5ff854b59f1aa1841b505281aaeabd7d/packages/e2e-tests/specs/editor/plugins/__snapshots__/cpt-locking.test.js.snap#L43

Looks like we can add attributes here if required: https://github.com/WordPress/gutenberg/blob/d70856b4b0cea364231c0981ada552723a3bb4c1/packages/e2e-tests/plugins/cpt-locking.php#L134

ramonjd committed:
Clicking on block variation skip

ramonjd committed:
Programmatically creating a group block will not default to the constrained layout type. It must be specified.

ramonjd commented:
> Check whether the templates need a layout<!-- wp:group {"templateLock":false,"layout":{"type":"constrained"}} -->. 

With this PR, if a group block is programmatically created without a layout (e.g., with `registerBlockType`, `register_post_type` [docs](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-templates/#api)), and the template doesn't set a default layout type of 'constrained', children of a wide-aligned group block will not "constrain" themselves within the content width.

<img width="1368" alt="Screen Shot 2022-08-25 at 2 19 10 pm" src="https://user-images.githubusercontent.com/6458278/186574057-c208a951-3804-4452-a7c8-5c0757d9d85e.png">

<details>

<summary>Example block HTML</summary>

```html
<!-- wp:group {"align":"full","backgroundColor":"vivid-red"} -->
<div class="wp-block-group alignfull has-vivid-red-background-color has-background"><!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"align":"right"} -->
<p class="has-text-align-right">Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"},"backgroundColor":"luminous-vivid-orange"} -->
<div class="wp-block-group has-luminous-vivid-orange-background-color has-background"><!-- wp:paragraph -->
<p>Paragraph inside a Group block inside a Group block</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->

<!-- wp:group {"layout":{"type":"constrained"},"align":"full","backgroundColor":"vivid-purple"} -->
<div class="wp-block-group alignfull has-vivid-purple-background-color has-background"><!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"align":"left"} -->
<p class="has-text-align-left">Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"align":"right"} -->
<p class="has-text-align-right">Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Paragraph inside a Group block</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"},"backgroundColor":"luminous-vivid-orange"} -->
<div class="wp-block-group has-luminous-vivid-orange-background-color has-background"><!-- wp:paragraph -->
<p>Paragraph inside a Group block inside a Group block</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->

```

</details>


This is a bit of a blocker in my opinion, at least until we work out an acceptable way around it. 


The placeholder is premised on the ability to select a layout with the implication that no layout is yet selected. [A recent PR](https://github.com/WordPress/gutenberg/pull/42763) introduced changes the Group block so **newly-added blocks** are automatically set to “constrained” layout. Newly-added comprises blocks added in the editor and those programmatically, via templates.

If we are to honour the default layout attributes in the group block, we need to somehow know that a block has been recently inserted in the editor, since all newly-created blocks will have the default layout.  `wasBlockJustInserted( clientId )` does not work. 

An alternative is to add a flag in the block json to indicate that the default value is being used:

```json
		"layout": {
			"type": "object",
			"default": {
				"type": "constrained",
				"isDefault": true
			}
		}
```

Then, in `edit.js`, we know to show to the placeholder in the editor if this attribute prop exists, and there are no inner blocks too. When a use selects a layout `isDefault` won't exist.

```
	const { tagName: TagName = 'div', templateLock, layout = {} } = attributes;
	const { type: layoutType = null, isDefault = false } = layout;

	// Whether to show the variations placeholder.
	// `isDefault: true` only exists in the default layout attributes in block.json
	// in order to identify blocks that have been inserted, programmatically or otherwise, with no changes.
	// When a user selects a layout `isDefault` won't appear in the block's attributes.
	const showPlaceholder = isDefault && ! hasInnerBlocks;
```

ramonjd committed:
Reinstating default "constrained" layout type, but adding an extra ephemeral property to identify blocks that have been inserted.

ramonjd commented:
Another bug:

When inserting a Group block, the block is not focussed as with all other blocks. 

Tried `selectBlock` after insertion. 

This test fails: https://github.com/WordPress/gutenberg/blob/d6a073daa3279bdc55c961e644c4fddc3f09c53d/packages/e2e-tests/specs/editor/various/keyboard-navigable-blocks.test.js#L230

ramonjd committed:
Ensure that the inserted block is selected after a Group variation is selected.

ramonjd commented:
I've addressed the default 'is-constrained' in https://github.com/WordPress/gutenberg/pull/43496#issuecomment-1226756445 and selected block issues in https://github.com/WordPress/gutenberg/pull/43496#issuecomment-1226803315 in [c1b2044](https://github.com/WordPress/gutenberg/pull/43496/commits/c1b204438af8e3cac98caa675ad05208e8feacdd) and [d5ef0f8](https://github.com/WordPress/gutenberg/pull/43496/commits/d5ef0f8c8daba30f9ee3210f0ee6781b2949cc75) respectively.

jasmussen commented:
Nice, thanks for the ping! From #43433, which this would fix, here's a mockup for how this could look:

<img src="https://user-images.githubusercontent.com/1204802/185644454-cef54532-21a9-4ca1-b798-6baa9690bca7.png">

Here's the current state of the branch:

<img width="672" alt="Screenshot 2022-09-06 at 08 39 52" src="https://user-images.githubusercontent.com/1204802/188564324-d7ddda2a-ac0e-4061-a9a5-95cb6fc7c0ca.png">

I'm happy to help push style changes here if helpful, but for now I'll share changes as a list, then I trust you to let me know if my assistance would be helpful:

* We don't need the "Skip", as it does the same as clicking Group.
* Anything we can do to move away from the classic placeholder style (icon/title/description) for this one, and towards the leaner one would be good, as we need for placeholders to contain as little content as possible to scale to as many narrow contexts as possible.
* Should use the new description.
* Should use tertiary style of buttons, with new illustrations which I'll provide the SVGs for in a moment.

No need to add the following illustrations to the icon library, these can exist just for the Group. Stack:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 44 32" style="enable-background:new 0 0 44 32" xml:space="preserve"><path d="M42 0H2C.9 0 0 .9 0 2v12.5c0 .6.4 1 1 1h42c.6 0 1-.4 1-1V2c0-1.1-.9-2-2-2zm1 16.5H1c-.6 0-1 .4-1 1V30c0 1.1.9 2 2 2h40c1.1 0 2-.9 2-2V17.5c0-.6-.4-1-1-1z" style="fill:#ccc"/></svg>
```

Row:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 44 32" style="enable-background:new 0 0 44 32" xml:space="preserve"><path d="M42 0H23.5c-.6 0-1 .4-1 1v30c0 .6.4 1 1 1H42c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2zM20.5 0H2C.9 0 0 .9 0 2v28c0 1.1.9 2 2 2h18.5c.6 0 1-.4 1-1V1c0-.6-.4-1-1-1z" style="fill:#ccc"/></svg>
```

Group:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 44 32" style="enable-background:new 0 0 44 32" xml:space="preserve"><path d="M42 0H2C.9 0 0 .9 0 2v28c0 1.1.9 2 2 2h40c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2z" style="fill:#ccc"/></svg>
```

Let me know what you think!

jasmussen commented:
> Anything we can do to move away from the classic placeholder style (icon/title/description) for this one, and towards the leaner one would be good, as we need for placeholders to contain as little content as possible to scale to as many narrow contexts as possible.

To elaborate a bit, I would expect every classic placeholder to eventually move towards either the dashed style (see unselected image placeholder) or the above style (white card but with minimal setup options). Columns and Table also come to mind. This is not something that needs to happen in short order for all of those, but is nevertheless a destination to move towards, and worth considering how that impacts the implementation, i.e. the above isn't a one-off for the group.

ramonjd commented:
Thanks for the guidance, @jasmussen! 🙇 

I was hoping that we could perhaps get the base functionality in and then work on the styles immediately after merging? It might be easier to iterate and separate the code review from the style (?) Also to keep the size of this one in check.

There are a couple of things that I think wouldn't create too much noise among the refactoring etc going on this PR however.

> We don't need the "Skip", as it does the same as clicking Group.
> Should use the new description.

Definitely doable. I will commit the following changes:

<img width="400" alt="Screen Shot 2022-09-06 at 8 35 55 pm" src="https://user-images.githubusercontent.com/6458278/188614274-af87482d-0d56-4cf0-9eeb-4bc30fb70b4a.png">

_Edit_: the skip button features in a lot of E2E tests so I'll leave that for another commit so I don't break CI 🤭 

> Anything we can do to move away from the classic placeholder style (icon/title/description) for this one, and towards the leaner one would be good, as we need for placeholders to contain as little content as possible to scale to as many narrow contexts as possible.

I've reused the `<__experimentalBlockVariationPicker />`, which is also used by the [Columns block](https://github.com/WordPress/gutenberg/blob/29d6801902ad97f749ccadd789a23b2d97369e3e/packages/block-library/src/columns/edit.js#L226-L225). So we could probably refactor/redesign that one at the same time?

<img width="400" alt="Screen Shot 2022-09-06 at 8 12 09 pm" src="https://user-images.githubusercontent.com/6458278/188609300-4be8d06c-2500-4c32-a76f-ad0984024bf6.png">

> Should use tertiary style of buttons, with new illustrations which I'll provide the SVGs for in a moment.

If we want to create a reusable component with the leaner style might require a bit of a refactor, and/or an extension of the variations objects, perhaps adding a `placeholderIcon` field for an alternative SVG. 

React also complains about the `style` attribute in the svg and path elements so we might have to save them as images.

I did a quick test of just swapping the icons and removing the offending svg props:

<img width="400" alt="Screen Shot 2022-09-06 at 8 30 23 pm" src="https://user-images.githubusercontent.com/6458278/188615680-c78daa06-1cf4-45cd-86d2-3642e0c6f26a.png">

ramonjd committed:
Removing skip link and update description

jasmussen commented:
In order to land this before the freeze, yes we probably can work towards a minimal v1 and then move fast on the visual changes after that fact. 

We should distinguish between icons and illustrations. The group/row/stack icons should not change (and they also shouldn't be scaled up to double size). Columns is an example as it retains its default columns icon, but includes illustrations for the various variants in `columns/variations.js`. We can definitely do the same here, for starters with the SVGs provided, but depending on how far we get in the followups we could tweak the group variation illustration style to be outlined like the columns ones are. I'd prefer we improve the Columns variations instead, though, and I'd be happy to provide the new illustration style icons to those variations instead.

What do you think?

ramonjd commented:
> We should distinguish between icons and illustrations. The group/row/stack icons should not change (and they also shouldn't be scaled up to double size). 

Definitely should be the goal 👍 

> Columns is an example as it retains its default columns icon, but includes illustrations for the various variants in columns/variations.js. We can definitely do the same here, for starters with the SVGs provided, but depending on how far we get in the followups we could tweak the group variation illustration style to be outlined like the columns ones are.

The difference between columns and group variations is that the group variations are shown in the inserter and transform scopes, and there's not yet a way to differentiate between scopes for the purpose of distinguishing between icons and illustrations (or anything else).

So my guess is one would have to extend the variations API, or create something bespoke for variation placeholders. Not sure yet. 

It makes me think that it might take a bit longer than is available before freeze to get something stable through the hoop, on top of the other migration work. 🤔 

If the functionality as it is appears like an improvement, it could be part of 6.1, then work could continue on both the UI and the plan to rework other placeholders later?

ramonjd committed:
Reinstate skip for now until we can update the e2e tests

jasmussen commented:
With improvements to group/row/stacks on their own (clearer icons, multiple interfaces for transforming, clearer empty resting states), #43433 does need a bit of fidelity in order to be a better experience. I do think the right design helps it achieve that, but would also agree, if we don't think we can get there, maybe it won't be a 6.1 thing, but it would still benefit the plugin. So we have until sep 20 to figure that out. Is the main problem the need to refactor the placeholder component or the variation picker?

ramonjd commented:
> So we have until sep 20 to figure that out. Is the main problem the need to refactor the placeholder component or the variation picker?

Good question!

I think the answer is "potentially both" but maybe only the variation picker 😄 

We'd need to work on the [variation picker](https://github.com/WordPress/gutenberg/tree/cd0f37a1050c0a763e0547b5057c68164dbfaf32/packages/block-editor/src/components/block-variation-picker) both in terms of styling and code - positioning the variations, and adding logic to use images/icons other than those used in the inserter for example.

This way we could take on the columns and group placeholders at the same time, if that's desirable.

The variation picker uses the placeholder component - I'm not 100% sure we'd need to do anything quite yet, unless we want to, for example, change the outline of all placeholder containers.

Ultimately, I don't think it'll be difficult task to tweak the design, yet believe it'd be better suited to separate PRs as they would be easier to review and test.

I'm very happy to work on it, but the catch is I'm away next week and still have to devote some time to migrating all the PHP files it appears I've touched this year 🤣

ramonjd commented:
I've been playing around to test some of my assumptions.

For this PR (or a direct follow up)[^1], I can get it looking something like this without too much refactoring ping pong and PR clutter:

https://user-images.githubusercontent.com/6458278/188777007-de7882db-aebe-4120-b853-d3c2ae21b7cd.mp4




I'm still convinced that it'd be better to tackle the placeholder/variation placeholder holistically. For example, the main "Group" heading is controlled by `<Placeholder />` so it'd need an exception. Also, the labels beneath the button would need an exception. 

[^1]: I've pushed the change to this branch. We can roll back if we want to merge functionality now and design later.

ramonjd committed:
Stylising the placeholder icons
updating tests

tellthemachines commented:
> I've pushed the change to this branch. We can roll back if we want to merge functionality now and design later.

Is there any downside to merging this design now? If we're changing the UI ideally we should do it in as few steps as possible, so as not to subject users to constant change. 

I haven't looked at the code in detail but it's working pretty well in testing ✅

jasmussen commented:
Thanks so much for looking!

I'm also a big fan of small iterations, but we want to be sure that we can get [very close to the mockup here](https://github.com/WordPress/gutenberg/pull/43496#issuecomment-1237725622), for 6.1, otherwise I would lean towards us doing this in, as you call it, "the holistic way". 

To put it differently: this one is important in the slightly longer than 6.1 view, but it isn't as urgent as everything else is, so we don't have to rush it.

ramonjd commented:
> To put it differently: this one is important in the slightly longer than 6.1 view, but it isn't as urgent as everything else is, so we don't have to rush it.

Good to know, thanks a lot! @jasmussen 🙇 

I guess we can put this PR in the fridge for a while then. I'm not sure when I can return to it, but I'll try to keep it rebased.

To summarize, I'd recommend:

- refactoring the Placeholder and  [variation picker](https://github.com/WordPress/gutenberg/tree/cd0f37a1050c0a763e0547b5057c68164dbfaf32/packages/block-editor/src/components/block-variation-picker) components to accommodate the [mockup designs](https://github.com/WordPress/gutenberg/pull/43496#issuecomment-1237725622)
- Come up with a better way to have distinct "placeholder icons", well, better than what this PR is doing 😆 
- identify other areas that may be affected by such work, e.g., the columns variation picker, check for regressions etc

ramonjd commented:
This might need a new approach after https://github.com/WordPress/gutenberg/pull/44176, which removes the default attributes in block.json.

tellthemachines commented:
> This might need a new approach after https://github.com/WordPress/gutenberg/pull/44176, which removes the default attributes in block.json.

I haven't looked into the fine detail, but if you're using the block variations to populate the placeholder, that part should Just Work. The constrained layout variation is now set to default, which should have the same effect as the bit that was removed from `block.json`.

ramonjd commented:
> I haven't looked into the fine detail, but if you're using the block variations to populate the placeholder, that part should Just Work. The constrained layout variation is now set to default, which should have the same effect as the bit that was removed from block.json.

Thanks for chiming in! 

I haven't spent a great deal of time thinking about it either, but the challenge after rebase for me was to build the condition that determined the Placeholder visibility.

tellthemachines commented:
> I haven't spent a great deal of time thinking about it either, but the challenge after rebase for me was to build the condition that determined the Placeholder visibility.

Oh I see, that might be tricky. Could we look at both the attributes and the content, and only show the placeholder if there are no non-defaults set _and_ no inner blocks?

ramonjd commented:
> Could we look at both the attributes and the content, and only show the placeholder if there are no non-defaults set and no inner blocks?

For sure! That's the approach I took in the current iteration, by adding `isDefault` to the default block.json layout attributes. 

It was stripped when replaced by the variant's attributes after selection.

I might try to see if I can shoehorn it in again, since relying on available defaults, e.g., 

```js
const showPlaceholder = 'constrained' === layoutType && 'div' === TagName && ! hasInnerBlocks;
```

gives this effect:

![2022-09-21 12 19 54](https://user-images.githubusercontent.com/6458278/191400474-c208b41d-00c4-419e-8dd0-f08876c1cf51.gif)

ramonjd committed:
Adding `isDefault` to the default props and removing it when a group variation is selected.

ramonjd committed:
Updating tests to account for:
- default layout is "constrained" when programmatically inserting a group block
- using createBlock with no attributes passed

ntsekouras commented:
Thanks for your work here @ramonjd!

I was thinking about this issue and I believe we shouldn't use `__experimentalBlockVariationPicker` and even if we did, we should be cautious about making changes to that component, as is being reused elsewhere. For example here we are adding `variation.placeHolderIcon` which is not part of the variations API..

What I believe is we should add a new component specifically for Group block, but have in the back of our minds that it could be someday the new placeholder picker and could be extracted to a reusable component.

ramonjd commented:
> What I believe is we should add a new component specifically for Group block, but have in the back of our minds that it could be someday the new placeholder picker and could be extracted to a reusable component.

This is easy enough to change, thanks for your thoughts @ntsekouras!

There will probably be further customizations required as it evolves into a "layout picker", so I agree. Not sure when I'll get back to this, but I'll move over to a new component in the next commit.

ramonjd committed:
Creating a custom group block variation picker so we do not have to update (and potentially break) the blockvariationpicker component, which is used elsewhere.

ramonjd committed:
Add missing button label.
Add class to button.
Fix outline heights for button.

ramonjd commented on packages/block-library/src/group/variations.js:
```diff
@@ -9,9 +9,9 @@ const variations = [
 		title: __( 'Group' ),
 		description: __( 'Gather blocks in a container.' ),
-		attributes: { layout: { type: 'constrained' } },
+		attributes: { layout: { type: 'constrained', isDefault: true } },
```
`isDefault: true` is the magic flag that allows us to know whether a block is in its "default" state, that is, recently inserted with no layout selected. 

I'm using it in combination with `hasInnerBlocks` to know when to show the placeholder.

E.g., `const showPlaceholder = isDefault && ! hasInnerBlocks;`.

When a layout is selected in the editor, the `isDefault` attribute is stripped so it doesn't appear in the markup, at least if added in the editor.

I'm not yet confident this is the right way to go about things as it adds an attribute by default to group block that meaningless outside the context of the editor. 

Maybe there's another way, or maybe what I'm doing is fine. 

I'll keep testing and try to come up with something.

I'll try to think of an alternative.

youknowriad replied:
That seems a bit fragile to me because if unchanged, that value is going to persist in the post. Why are we not showing the placeholder state for all group blocks (not row or stack) without any inner blocks?

ntsekouras replied:
I agree with Riad. I thought an alternative would be to just check if it has innerBlocks and apply some classes, instead of [only-child styles](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/group/editor.scss#L27), when we have selected something in the placeholder and still don't have any innerBlocks. This way we would have the `revert to placeholder` behavior discussed [here](https://github.com/WordPress/gutenberg/issues/43433#issuecomment-1257872806):

>If you just click "Group" in the flow illustrated, you should get the group as it exists today, i.e. with the inserter. If you then don't add any blocks inside and deselect, we can revert back to the setup state as shown. Does that make sense?

I haven't tried it out though..

ramonjd replied:
Thanks for the feedback. My instinct was warning me something wasn't right, so I appreciate the confirmation, folks.

> Why are we not showing the placeholder state for all group blocks (not row or stack) without any inner blocks?

What I was attempting to avoid, perhaps unnecessarily and despite @jasmussen's [comments](https://github.com/WordPress/gutenberg/issues/43433#issuecomment-1257872806), was the situation in which a user:

1. Inserts a group block
2. Selects a variation, but does not add any child blocks
3. Works a bit, saves.
4. Returns to the page to find a group variation placeholder
5. Wonders why they have to select a variation again

I'll remove the `isDefault` attribute so folks can test. Perhaps I was, as usual, overthinking it.

ramonjd replied:
> check if it has innerBlocks and apply some classes, instead of only-child styles, when we have selected something in the placeholder and still don't have any innerBlocks

Do you mean create a new classname to indicate that a variation has been selected, regardless of whether there are innerBlocks or not?

That would be an alternative to the `isDefault` flag. It would also create a persistent value. 

It's worth pursuing. I'll try it out. Thanks!

🍺

ramonjd replied:
> What I was attempting to avoid, perhaps unnecessarily and despite @jasmussen's https://github.com/WordPress/gutenberg/issues/43433#issuecomment-1257872806, was the situation in which a user:

I withdraw that. This is the assumption I was going on, from the [comments](https://github.com/WordPress/gutenberg/pull/43496#issuecomment-1298249501) below:

> Since this new state is mainly intended to be shown for an empty group inserted fresh,

tellthemachines replied:
> Why are we not showing the placeholder state for all group blocks (not row or stack) without any inner blocks?

If we show the placeholder for Groups without inner blocks, or even without a variation selected, we'll lose the possibility of using Groups for visual/aesthetic purposes, such as [this](https://block-museum.com/2022/01/26/026/). We need to check for the existence of any custom styling before showing the placeholder.

ramonjd replied:
> If we show the placeholder for Groups without inner blocks, or even without a variation selected, we'll lose the possibility of using Groups for visual/aesthetic purposes, such as [this](https://block-museum.com/2022/01/26/026/). We need to check for the existence of any custom styling before showing the placeholder.

Also a great line of reasoning, thanks @tellthemachines 

Maybe a utility class or other HTML attribute is called for – `is-empty` or `aria-placeholder,="Empty Group block"` – to replace the `isDefault` block attribute and work in conjunction with the placeholder as a flag to show that a variation hasn't been selected.

I'll experiment and seek feedback.

ramonjd replied:
> Why are we not showing the placeholder state for all group blocks (not row or stack) without any inner blocks?

Also, I forgot say, the default state of a Group block has no inner blocks. So even if a user selects the default variation, we need something else to tell the editor to remove the placeholder since it'll still have no inner blocks.

jasmussen commented:
Let me know if/when you need a fresh review. Thanks for working on this!

ramonjd committed:
Update native test fixture

Updating design/spacing of placeholder
Hardcoding custom icons

ramonjd commented:
> Let me know if/when you need a fresh review. 

Thank you! Any time you get to cast an eye on it will help. 🙇

jasmussen commented:
Took a brief look at the current state:

![state](https://user-images.githubusercontent.com/1204802/198243815-635aeef0-98aa-4bed-a26e-728b06672879.gif)

Nice, the behavior looks just about right in terms of the flow. I imagine lots of little things are still in flux, but I'll list out a couple of todos in order to get to the target design:

<img width="784" alt="Screenshot 2022-10-27 at 11 13 06" src="https://user-images.githubusercontent.com/1204802/198244222-e932f131-2645-464f-a0bc-0d7632e39c41.png">

Hopefully that can give an opportunity to surface the challenges we need to work through to get there:

* Icons/buttons should be exactly 44x32 with 0px padding, right now they are 64x64 with 6px padding.
* There should be a 24px gap between buttons, right now there are 16px margins left and right of each button coming to 32px. We can either use proper `gap`, or those left/right margins should be 12px. 
* The space between the description and the row of buttons should be 18px, and the inner padding of the entire container should be 24px. 
* There should be no "Group" icon and label in the placeholder.

Here's a screenshot that may better highlight some of the metrics:

<img width="596" alt="Screenshot 2022-10-27 at 11 17 55" src="https://user-images.githubusercontent.com/1204802/198245203-7a118253-f47f-4916-a43d-adc4a14ddcd5.png">

If it helps, we can think of this as an entirely new placeholder component to slowly replace every other similar placeholder, such as that of columns:

<img width="726" alt="Screenshot 2022-10-27 at 11 19 33" src="https://user-images.githubusercontent.com/1204802/198245573-132005a6-1a48-416e-8123-bebc434735e0.png">

ramonjd commented:
Thanks for keeping this one alive @jasmussen 

I've tweaked the CSS a bit:

https://user-images.githubusercontent.com/6458278/199156899-4039cfc8-13dc-4dd3-9e8a-ce285d66a637.mp4


One question remains: do you think we should hide the instructions for medium and small placeholders? Here they are displayed:

<img width="911" alt="Screen Shot 2022-11-01 at 3 08 37 pm" src="https://user-images.githubusercontent.com/6458278/199156852-3db9e6ef-2093-49d2-b6d1-f3fa44d17a3c.png">

jasmussen commented:
Nice! Fast work!

> One question remains: do you think we should hide the instructions for medium and small placeholders? 

That's an excellent question. Since this new state is mainly intended to be shown for _an empty group inserted fresh_, it could make sense to jump straight to the 2nd group state (showing the plus) in nested contexts.

However it's probably best to start without this change, just to keep the PR simple, especially if it's an easy followup. That also affords us a chance to evaluate the change. What do you think?

ramonjd commented:
> However it's probably best to start without this change, just to keep the PR simple, especially if it's an easy followup. That also affords us a chance to evaluate the change. What do you think?

No worries. Make good sense. If we think we need the instructions title globally, it's no dramas to add it in. Cheers!

ramonjd committed:
Removing the `isDefault` Group block attribute. It's a fragile solution to keep track of whether a variation has been selected since it adds a potentially useless attribute.
In this iteration, we'll keep track of whether to show the placeholder via useState in the editor, and, more permanently, whether the block has innerblocks.

ramonjd commented:
Following on from this [conversation](https://github.com/WordPress/gutenberg/pull/43496#pullrequestreview-1155480419), I've updated this PR with the following:

- Setting the placeholder state in the editor using `useState`, with the default value being `! innerBlocks`. This means that the variation placeholder will always show on a reload if the block has no inner blocks.
- Styled blocks with no innerBlocks still show the placeholder. The styles are applied to the placeholder container however. Taking into account all possible current and future style attribute AND style object values, I think there are too many variables to ascertain for sure whether a block has style modifications. If there's a cheap way at the moment, I'd love to know it.

https://user-images.githubusercontent.com/6458278/199380030-4c490b7c-a90d-42cb-a34a-b20a1a48ce52.mp4

As for style blocks, they'll appear as expected on the frontend:

<img width="500" alt="Screen Shot 2022-11-02 at 1 14 05 pm" src="https://user-images.githubusercontent.com/6458278/199380086-9136a752-04bf-452b-9424-993e118dc2e1.png">

I think this is a decent first iteration. 

Another potential approach could be to somehow change the `role` attribute of the block to `role="presentation"` until it has innerBlocks, and use that as a hook.

ramonjd committed:
Ensure that we select the default group variation in the e2e tests.

jasmussen commented:
This is remarkably close to the design, kudos:

![group](https://user-images.githubusercontent.com/1204802/199428087-8004058a-18cb-4efe-b1b4-0dec3e5a1d0b.gif)

How do you feel about the CSS and structure, in terms of this becoming a template for new "starting point blocks"? Does it scale well?

tellthemachines commented on packages/block-library/src/group/edit.js:
```diff
@@ -52,14 +88,21 @@ function GroupEdit( {
 	const { type = 'default' } = usedLayout;
 	const layoutSupportEnabled = themeSupportsLayout || type === 'flex';
+	const [ showPlaceholder, setShowPlaceholder ] = useState(
+		! hasInnerBlocks
```
If we grab `style` from attributes and make this initial state `! hasInnerBlocks && ! style` we should cover most cases of visual effects Groups, because the majority of color, dimension and border settings are inside the `style` object. The only one that doesn't seem to be is border color, which feels like a bug? Layout also isn't but for styling purposes it's irrelevant.

ramonjd replied:
I did try this out, but it seemed inconsistent with behaviour when setting presets.

What do you think about checking for style and also the two, currently-available presets?

 `! hasInnerBlocks && ! style && ! attributes?. backgroundColor && ! attributes?. fontSize`

The catch is that if we add presets for, say, spacing (which [I think is on the cards](https://github.com/WordPress/gutenberg/issues/45431)), we'll have to update the condition.

ramonjd replied:
Oh, and `textColor` 😄

tellthemachines replied:
Oooh of course, the presets are outside `style` 🤦 
Still, the amount of presets is limited, and unlikely to increase by much so I think it's worth the tradeoff to have a better experience.

tellthemachines commented on packages/e2e-tests/specs/editor/plugins/__snapshots__/cpt-locking.test.js.snap:
```diff
@@ -41,7 +41,7 @@ exports[`cpt locking template_lock all should not error when deleting the cotent
 
 exports[`cpt locking template_lock all unlocked group should allow blocks to be moved 1`] = `
-"<!-- wp:group {\\"templateLock\\":false} -->
+"<!-- wp:group {\\"templateLock\\":false,\\"layout\\":{\\"type\\":\\"constrained\\"}} -->
```
Oh interesting, so this changed again without the test being altered in any way?

ramonjd replied:
There was a related change above: https://github.com/WordPress/gutenberg/pull/43496/files#diff-8ea13eb3838ceaa967fbed75698db9216cea3d86ebd584423730824556c7b974R72

Thanks for pointing it out however. Maybe we don't need the change anymore.

tellthemachines reviewed (commented):
Code looks good and it's working well in testing! Just a couple of comments below and one thought:

It feels irksome to have the placeholder appear when we deliberately select a Row or Stack block from the inserter. Could we perhaps check for `usedLayout.type !== 'flex'` in the conditions for displaying it?

ramonjd commented:
> It feels irksome to have the placeholder appear when we deliberately select a Row or Stack block from the inserter. Could we perhaps check for usedLayout.type !== 'flex' in the conditions for displaying it?

This sounds like a good compromise. Thanks for the idea, and for testing again @tellthemachines . I'll test it out!

ramonjd commented:
> How do you feel about the CSS and structure, in terms of this becoming a template for new "starting point blocks"? Does it scale well?

There's plenty of scope for abstraction. These changes are coupled tightly to the Group block, and it uses the Placeholder component, which is useful because it tracks resizes via `is-small` etc

I think it's a good start. If you're happy with the way things look, the next step after this PR would be to abstract the component and allow custom icons as props and so on. 

I threw a few more variations in there to test and things stack okay, though maybe we could look at grid during abstraction.

https://user-images.githubusercontent.com/6458278/199647936-7361f5d3-e069-4cda-9874-de32ed752228.mp4

ramonjd committed:
Creating custom hook for `showPlaceholder`
Testing on the CI whether the changes to CPT are required :)

ramonjd commented:
I've pushed a change to test @tellthemachines's [suggestions above](https://github.com/WordPress/gutenberg/pull/43496#discussion_r1012411145).

I create a new hook `useShouldShowPlaceHolder`. If it's okay I'll create some tests for it, fix the failing e2e tests 😄  and we should be right to rock.

ramonjd committed:
Default selected layout type of Group variation is "constrained"

ramonjd committed:
Am I pretty enough for you, linter?

ramonjd committed:
Reverted

tellthemachines reviewed (approved):
Thanks for the iteration! Changes are looking good and testing as expected in both post and site editor ✅

aristath reviewed (approved):
LGTM! 👍

paaljoachim commented:
Testing.
Downloading the Gutenberg PR build.
Twenty Twenty Three.
Using the Site Editor.

### Before:
![Before Group block placeholder](https://user-images.githubusercontent.com/5323259/200521831-dc4f745b-453c-4ff9-831b-53a75efe1fba.jpg)

### After:
![After Group block placeholder](https://user-images.githubusercontent.com/5323259/200521883-3f7970e1-5513-49b5-b73c-3356772f96e8.jpg)

Because of consistency and recognition I would expect to see the sidebar container icons also in the placeholder like so.
![After Group block placeholder other icons](https://user-images.githubusercontent.com/5323259/200522649-fd5865e4-64ed-4d94-9d9c-a14b95930a94.jpg)

OR

That the sidebar icons are changed like so.

![After Group block placeholder change sidebar icons](https://user-images.githubusercontent.com/5323259/200522893-9685bd22-7130-45e7-a025-d6d6c44bfb2b.jpg)

The placeholder is the first place where one meets the initial settings for any given block. As the placeholder is briefly there until the user has made the first choice. It is natural to strengthen what one initially sees also in the sidebar. Be it in the Group block, Columns block or other blocks.

ramonjd committed:
Reinstating `useEffect` so that we can react to changes in the style state.

ramonjd commented:
Thanks for testing folks. I just had to (re)add a `useEffect` in the hook so that the group variation picker reacts to styles changes, i.e., disappears. It's just repetitive logic, so I think we're still okay to merge.

paaljoachim commented:
Are there plans to change either the placeholder icons or the Gather block in a container icons?
As these do the same thing, right?

If they do the exact same thing it would be beneficial to keep the icons the same. As the same feature should also look the same. This is kind of tricky as both sets of icons have their advantages and disadvantages. 

The new icons are easier to understand based on the visual representation we see.
The existing Gather block in a container icons are more "fancy" not as easy to understand but these are now established into Gutenberg.

 @jasmussen

jasmussen commented:
> Are there plans to change either the placeholder icons or the Gather block in a container icons?
> As these do the same thing, right?

No, the icons in the inspector are quick transform shortcuts. The illustraitons in the placeholder are starting points, and while at the moment they appear to do the same, I could see the starting points expanding to have more than 3 options. Perchance there could be a starting point that was more equivalent to a wireframe block pattern, and fast-forward you to a row with stack blocks inside, for a flex based column setup. See also https://github.com/WordPress/gutenberg/pull/43496#issuecomment-1293237708

I think this one is good to go as is, and we can follow up!
